### PR TITLE
Add basic phpmyadmin implementation

### PR DESCRIPTION
**Pull request by mohe2015**
###### Motivation for this change
phpmyadmin is useful to quickly do database changes or backups / recoveries.

###### Things done
Added phpmyadmin package.
Added phpmyadmin module.

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aanderse** [comment]
A couple initial thoughts without having given this a thorough review:
- I'm always shocked to see new modules using `httpd` instead of `nginx` - what was your rationale here?
- `phpmyadmin` security can be questionable and updates are **essential** to get out on time; what is your plan and commitment level to this package?
- configuring `phpmyadmin` should usually be pretty customized so can your module adequately satisfy these needs while adding an advantage over the sysadmin installing/configuring `phpmyadmin` on their own without a module?

**Comment by mohe2015** [comment]
> I'm always shocked to see new modules using `httpd` instead of `nginx` - what was your rationale here?

wordpress module is written for httpd and I'm still using httpd. I didn't want to run two webservers. An abstraction layer seems to be really needed but I won't be able to write one. I won't rewrite this but I'm fine if this is not merged.

> `phpmyadmin` security can be questionable and updates are **essential** to get out on time; what is your plan and commitment level to this package?

I'm watching the phpmyadmin repo and should be able to provide timely updates.

> configuring `phpmyadmin` should usually be pretty customized so can your module adequately satisfy these needs while adding an advantage over the sysadmin installing/configuring `phpmyadmin` on their own without a module?

I added code to configure phpmyadmins. This has the advantage of NixOS, namely being declarative.

**Comment by aanderse** [comment]
All reasonable answers, thanks. I'll try to get back to reviewing this as soon as I can.

File: nixos/modules/services/web-apps/phpmyadmin.nix [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        documentRoot = mkForce "${pkg hostName cfg}";
+        extraConfig = ''
+            <Directory "${pkg hostName cfg}">
+              <FilesMatch "\.php$">
+                <If "-f %{REQUEST_FILENAME}">
+                  SetHandler "proxy:unix:${config.services.phpfpm.pools."phpmyadmin-${hostName}".socket}|fcgi://localhost/"
+                </If>
+              </FilesMatch>
+              DirectoryIndex index.php
+              Require all granted
```
Comment (aanderse) [inline nixos/modules/services/web-apps/phpmyadmin.nix:141]: This line is scary.

File: pkgs/servers/web-apps/phpmyadmin/default.nix [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  ];
+
+  installPhase =
+  ''
+    mkdir -p $out
+    cp -r * $out/
+  '';
+
+  meta = {
+    homepage = https://www.phpmyadmin.net/;
```
Comment (aanderse) [inline pkgs/servers/web-apps/phpmyadmin/default.nix:23]: As of recently `homepage` needs to be quoted.

File: pkgs/servers/web-apps/phpmyadmin/default.nix [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    "installPhase"
+  ];
+
+  installPhase =
+  ''
+    mkdir -p $out
+    cp -r * $out/
+  '';
+
+  meta = {
```
Comment (aanderse) [inline pkgs/servers/web-apps/phpmyadmin/default.nix:22]: You're missing a few values in `meta` (`maintainer`, `platfotms`, see manual).

File: pkgs/servers/web-apps/phpmyadmin/default.nix [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    homepage = https://www.phpmyadmin.net/;
+    description = "phpMyAdmin is a free and open source administration tool for MySQL and MariaDB";
```
Comment (aanderse) [inline pkgs/servers/web-apps/phpmyadmin/default.nix:24]: `description` shouldn't start with the name of the package. See nixpkgs (not NixOS) manual for full details.

**Review by aanderse — CHANGES_REQUESTED** [review]
Sorry this took me so long to get back to. Let me know what you think.

File: nixos/modules/services/web-apps/phpmyadmin.nix [hunk 5]
```diff
@@ -0,0 +1,171 @@
+{ config, lib, pkgs, ... }:
+
+# heavily inspired by nixos/modules/services/web-apps/wordpress.nix
+
+let
+  inherit (lib) mkDefault mkEnableOption mkForce mkIf mkMerge mkOption types;
+  inherit (lib) any attrValues concatMapStringsSep flatten literalExample;
+  inherit (lib) mapAttrs mapAttrs' mapAttrsToList nameValuePair optional optionalAttrs optionalString;
+
+  eachInstance = config.services.phpmyadmin;
```
[the remark this answers is not in the document]
Comment (mohe2015) [reply nixos/modules/services/web-apps/phpmyadmin.nix:10]: I also thought about this and am still not sure about it. There may be usages where you have a lower security instance that is for one specific database (maybe for somebody else to access it) and a higher security instance that is additionally secured with basic authentication that can use a privileged database user. There is also the possibility to add other servers which you may not want on all instances. If you want any of the many settings in https://docs.phpmyadmin.net/en/latest/config.html to be different you need two instances.
Also it seems that you can [indeed](https://docs.phpmyadmin.net/en/latest/config.html#cfg_Servers_AllowDeny_rules) specify which user can connect.

So I personally would keep this although I didn't use multiple instances yet and also didn't really configure phpmyadmin to other settings than the default.

File: nixos/modules/services/web-apps/phpmyadmin.nix [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+  user = "phpmyadmin";
+  group = config.services.httpd.group;
+  stateDir = hostName: "/var/lib/phpmyadmin/${hostName}";
+
+  pkg = hostName: cfg: pkgs.stdenv.mkDerivation rec {
```
[the remark this answers is not in the document]
Comment (mohe2015) [reply nixos/modules/services/web-apps/phpmyadmin.nix:15]: Note to myself: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/web-apps/limesurvey.nix is also relevant.

**Review by mohe2015 — COMMENTED** [review]
I really appreciate the effort you put in the review and it also teached me a few things.

For the other changes I agree with you. I'm not sure when I will find time to implement them though as next week is my first week of studying Computer Science at an University. There are not too many changes if you disregard the "multiple instances" one so I hope soon.

**Comment by mohe2015** [comment]
@aanderse I think I addressed everything. I also added `extraHttpdConfig` to allow users to improve the security a lot. Let me know what you think.

**Comment by mohe2015** [comment]
@aanderse Friendly reminder if you're still interested in reviewing again or merging.

**Comment by aanderse** [comment]
Apologies. I'll try to find some time to prioritize this.

**Comment by mohe2015** [comment]
@aanderse new year, new luck ;-)

**Comment by mohe2015** [comment]
I will close this for now. https://www.adminer.org/ is also probably a more "modern" solution

**Pull request closed without merging**